src/osd/PrimaryLogPG.cc

```diff
@@ -13292,6 +13293,35 @@ int PrimaryLogPG::getattrs_maybe_cache(
   return r;
 }
 
+bool PrimaryLogPG::check_failsafe_full() {
+    return osd->check_failsafe_full();
+}
+
+void PrimaryLogPG::too_full(int32_t from)
+{
+  if (state_test(PG_STATE_RECOVERING)) {
+    //dout(10) << __func__ << ": Canceling recovery, out of space reported by "
+    //        << from << dendl;
+    queue_peering_event(
+          CephPeeringEvtRef(
+            std::make_shared<CephPeeringEvt>(
+              get_osdmap()->get_epoch(),
+              get_osdmap()->get_epoch(),
+              RecoveringTooFull())));
+  } else if (state_test(PG_STATE_BACKFILL)) {
+    //dout(10) << __func__ << ": Canceling backfill, out of space reported by "
+    //        << from << dendl;
+    queue_peering_event(
+      CephPeeringEvtRef(
+        std::make_shared<CephPeeringEvt>(
+          get_osdmap()->get_epoch(),
+          get_osdmap()->get_epoch(),
+          BackfillTooFull())));
+  } else {
+    assert(0);
+  }
+}
+
```

Comment:
This doesn't seem like enough.  We've already gotten the reservation.. what we really need is a way to abort recovery or backfill.  Until we have that figured out, I don't think this patch makes sense.